### PR TITLE
feat(core): enhanced address verification in sdk 

### PR DESCRIPTION
**Pull request by AbhayShekhawat94**
feat(core): enhanced address verification in sdk with optional request parameters forwarderVersion, baseAddress and allowSkipVerifyAddress flag for the createdAddress.

BG-39200

### CODE REVIEW

File: modules/core/src/v2/wallet.ts [hunk 1]
```diff
@@ -1341,19 +1359,23 @@ export class Wallet {
...
+        newAddress.baseAddress = !_.isUndefined(baseAddress) ? baseAddress : _.get(self._wallet, 'coinSpecific.baseAddress');
 
         const verificationData: VerifyAddressOptions = _.merge({}, newAddress, { rootAddress });
 
         if (verificationData.error) {
           throw new AddressGenerationError(verificationData.error);
         }
 
+        const impliedForwarderVersion = !_.isUndefined(addressParams.forwarderVersion) ? addressParams.forwarderVersion :
+        verificationData.coinSpecific?.forwarderVersion;
```
Comment (OttoAllmendinger) [inline modules/core/src/v2/wallet.ts:1371]: please add
```
/**
 * @prettier
 */
```
 at the top and format this file

**Review by OttoAllmendinger — CHANGES_REQUESTED** [review]

File: modules/core/src/v2/wallet.ts [hunk 2]
```diff
@@ -1257,6 +1259,8 @@ export class Wallet {
    * @param {Number} params.count=1 number of new addresses which should be created (maximum 250)
    * @param {Number} params.forwarderVersion The version of address to create, if applicable
    * @param {Boolean} params.lowPriority Ethereum-specific param to create address using low priority fee address
+   * @param {String} params.baseAddress base address of the wallet(optional parameter)
+   * @param {Boolean} params.strictMode set to strictly verify the created address
```
Comment (OttoAllmendinger) [inline modules/core/src/v2/wallet.ts:1263]: please explain the meaning of this in more detail (the condition in which we skip the verification)

File: modules/core/src/v2/wallet.ts [hunk 3]
```diff
@@ -1341,19 +1359,23 @@ export class Wallet {
         }
 
         newAddress.keychains = keychains;
-        newAddress.baseAddress = _.get(self._wallet, 'coinSpecific.baseAddress');
+        newAddress.baseAddress = !_.isUndefined(baseAddress) ? baseAddress : _.get(self._wallet, 'coinSpecific.baseAddress');
```
Comment (OttoAllmendinger) [inline modules/core/src/v2/wallet.ts:1362]: use `baseAddress ?? _.get(...)` instead

File: modules/core/src/v2/wallet.ts [hunk 4]
```diff
@@ -1341,19 +1359,23 @@ export class Wallet {
...
 
         const verificationData: VerifyAddressOptions = _.merge({}, newAddress, { rootAddress });
 
         if (verificationData.error) {
           throw new AddressGenerationError(verificationData.error);
         }
 
+        const impliedForwarderVersion = !_.isUndefined(addressParams.forwarderVersion) ? addressParams.forwarderVersion :
```
Comment (OttoAllmendinger) [inline modules/core/src/v2/wallet.ts:1370]: `addressParams.forwarderVersion ?? ...`

**Review by OttoAllmendinger — CHANGES_REQUESTED** [review]

File: modules/core/src/v2/wallet.ts (hunk 1 above)
Comment (AbhayShekhawat94) [reply modules/core/src/v2/wallet.ts]: Sure Otto, will add this once the file is ready to merge. Adding prettier does a lot of reformatting to the code which will make it really hard to review.

File: modules/core/src/v2/wallet.ts (hunk 2 above)
Comment (AbhayShekhawat94) [reply modules/core/src/v2/wallet.ts]: Done with the new commit.

File: modules/core/src/v2/wallet.ts (hunk 3 above)
Comment (AbhayShekhawat94) [reply modules/core/src/v2/wallet.ts]: Done

File: modules/core/src/v2/wallet.ts (hunk 4 above)
Comment (AbhayShekhawat94) [reply modules/core/src/v2/wallet.ts]: Done

**Review by OttoAllmendinger — APPROVED** [review]

File: modules/core/src/v2/wallet.ts (hunk 1 above)
Comment (OttoAllmendinger) [reply modules/core/src/v2/wallet.ts]: not a problem because we can review individually

**Review by OttoAllmendinger — APPROVED** [review]

File: modules/core/test/v2/unit/wallet.ts [hunk 5]
```diff
@@ -422,9 +434,238 @@ describe('V2 Wallet:', function () {
             forwarderVersion: 1,
           },
         });
-      await ethWallet.createAddress({ chain: 0, forwarderVersion: 1 }).should.be.rejectedWith('address validation failure: expected 0x32a226cda14e352a47bf4b1658648d8037736f80 but got 0x8c13cd0bb198858f628d5631ba4b2293fc08df49');
+       await ethWallet.createAddress({ chain: 0, forwarderVersion: 1 }).should.be.rejectedWith('address validation failure: expected 0x32a226cda14e352a47bf4b1658648d8037736f80 but got 0x8c13cd0bb198858f628d5631ba4b2293fc08df49');
     });
-  });
+
+    it('verify address when invalid baseAddress is passed', async function () {
+      nock(bgUrl)
```
Comment (tylerlevine) [inline modules/core/test/v2/unit/wallet.ts:441]: please store the return value and ensure the nock is consumed for all these nocks.

**Review by tylerlevine — CHANGES_REQUESTED** [review]
minor nit but please make sure you check all nocks to make sure they are actually consumed by your test. rest LGTM

File: modules/core/src/v2/wallet.ts (hunk 1 above)
Comment (AbhayShekhawat94) [reply modules/core/src/v2/wallet.ts]: Done

**Comment by AbhayShekhawat94** [comment]
> minor nit but please make sure you check all nocks to make sure they are actually consumed by your test. rest LGTM

Checked all tests individually and they are consuming the nocks properly.

File: modules/core/test/v2/unit/wallet.ts [hunk 6]
```diff
@@ -330,24 +330,55 @@ describe('V2 Wallet:', function () {
...
-          baseAddress: '0xdf07117705a9f8dc4c2a78de66b7f1797dba9d4e',
-        },
-        coin: 'teth',
-        keys: [
-          '598f606cd8fc24710d2ebad89dce86c2',
-          '598f606cc8e43aef09fcb785221d9dd2',
-          '5935d59cf660764331bafcade1855fd7',
-        ],
-      };
+    async function beforeEachSetCommonNocks() {
```
Comment (OttoAllmendinger) [inline modules/core/test/v2/unit/wallet.ts:349]: do this instead:

```

beforeEach('set common nocks', function () {
  nock(...)
});

afterEach(function () {
 nock.cleanAll();
});
```
});

**Review by OttoAllmendinger — CHANGES_REQUESTED** [review]

File: modules/core/test/v2/unit/wallet.ts (hunk 5 above)
Comment (OttoAllmendinger) [reply modules/core/test/v2/unit/wallet.ts]: this has not been resolved @AbhayShekhawat94

File: modules/core/test/v2/unit/wallet.ts (hunk 6 above)
Comment (OttoAllmendinger) [reply modules/core/test/v2/unit/wallet.ts:349]: and to satisfy tyler:

```
let nocks;
beforeEach('set common nocks', function () {
  nocks = [nock(...), nock(...)]
})

afterEach(function() {
  nocks.forEach(scope => scope.done());
  nocks.cleanAll();
});
```

File: modules/core/src/v2/wallet.ts [hunk 7]
```diff
@@ -1262,7 +1262,9 @@ export class Wallet {
    * @param {Number} params.forwarderVersion The version of address to create, if applicable
    * @param {Boolean} params.lowPriority Ethereum-specific param to create address using low priority fee address
    * @param {String} params.baseAddress base address of the wallet(optional parameter)
-   * @param {Boolean} params.strictMode set to strictly verify the created address
+   * @param {Boolean} params.strictMode set to strictly verify the created address.Throws error if address verification is skipped for any reason.
```
Comment (OttoAllmendinger) [inline modules/core/src/v2/wallet.ts:1343]: how about we call this `allowSkipValidation` instead? Default is `true` and we allow setting it to `false`

**Review by OttoAllmendinger — CHANGES_REQUESTED** [review]

File: modules/core/test/v2/unit/wallet.ts (hunk 5 above)
Comment (AbhayShekhawat94) [reply modules/core/test/v2/unit/wallet.ts]: Done

File: modules/core/test/v2/unit/wallet.ts (hunk 6 above)
Comment (AbhayShekhawat94) [reply modules/core/test/v2/unit/wallet.ts]: Done. Thanks Otto!

File: modules/core/src/v2/wallet.ts (hunk 7 above)
Comment (AbhayShekhawat94) [reply modules/core/src/v2/wallet.ts]: Done.

**Review by OttoAllmendinger — DISMISSED** [review]

File: modules/core/src/v2/baseCoin.ts [hunk 8]
```diff
@@ -60,6 +60,7 @@ export interface VerifyAddressOptions {
   }[];
   error?: string;
   coinSpecific?: AddressCoinSpecific;
+  impliedForwarderVersion?: number | undefined;
```
Comment (OttoAllmendinger) [inline modules/core/src/v2/baseCoin.ts:63]: either use `?` suffix or `| undefined`

**Review by OttoAllmendinger — APPROVED** [review]

**Review by OttoAllmendinger — DISMISSED** [review]

File: modules/core/src/v2/baseCoin.ts (hunk 8 above)
Comment (AbhayShekhawat94) [reply modules/core/src/v2/baseCoin.ts]: Done!

**Review by tylerlevine — APPROVED** [review]

**Pull request merged**